### PR TITLE
lib: A tutorial and some 'modify' monad rules for Lib.EquivValid

### PR DESCRIPTION
**Pull request by robs-cse**
Thanks to Toby Murray (@tobycmurray) for early feedback.

Signed-off-by: Robert Sison <robert.sison@unimelb.edu.au>

### CODE REVIEW

**Comment by robs-cse** [comment]
The tutorial can be built to PDF by running `./isabelle/bin/isabelle build -d . EVTutorial` from the l4v directory.

**Comment by lsf37** [comment]
Ok to ignore the link checker failures, these are hopefully temporary, and in any case unrelated. Also fine to ignore the `Refine` failures. They're running out of store and I should probably remove them from the test, because that happens too frequently now.

**Review by lsf37 — CHANGES_REQUESTED** [review]
The tutorial itself looks very nice from my side!

One small change request before I can approve: can you please add it to the regression test infrastructure, i.e. to the appropriate `tests.xml` and make sure it is also called in the corresponding github workflow. I think adding it to `.github/workflows/proof.yml `  in the run that calls `ASpec` and `AInvs` would make sense (this one also builds `Lib` as a dependency, IIRC).

File: lib/EVTutorial/ROOT [hunk 1]
```diff
@@ -0,0 +1,15 @@
+(*
+ * Copyright 2020, The University of Melbourne (ABN 84 002 705 224)
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ *)
+
+chapter Lib
+
+session EVTutorial = Lib +
+  options [document = pdf, document_output = "output"]
```
Comment (lsf37) [inline lib/EVTutorial/ROOT:10]: Unless you really want the user to inspect the generated sources, I would usually not add an `output` option.

```suggestion
  options [document = pdf]
```

**Comment by robs-cse** [comment]
The force-push to commit 1ee13dd is just a rebase, with some minor fixes to the last commit message.

**Review by mbrcknl — APPROVED** [review]
I've not read every detail of the tutorial, but the bits I looked at were great! I look forward to working through the rest...

Thanks for doing this!

**Review by lsf37 — APPROVED** [review]
👍

**Comment by robs-cse** [comment]
(Force-push to commit 3bd44af is another rebase.)

**Comment by robs-cse** [comment]
@lsf37 has advised that the two reviews are sufficient to approve the merge. The force push to commit e2194d4 is a rebase to the current mainline head, squashing all the fixup commits.

**Comment by lsf37** [comment]
Ok to ignore the failed link test. Apparently the nicta.com.au certificates have expired.

**Comment by robs-cse** [comment]
The force push to commit 19312b5 rebased to the mainline head, and passed [this testboard run](https://bamboo.ts.data61.csiro.au/browse/L4V-TESTBOARD-658). I'll "Rebase and Merge" it now - thank you again @lsf37 and @mbrcknl.

**Pull request merged**